### PR TITLE
OOP school library: set up associations 

### PR DESCRIPTION
**Pull request by Tracy-miranja**
- [ ] Create a class Classroom with the following:
- [ ] @label instance variable, should be initialized in the constructor.
- [ ] Setter and getter for @label (remember about attr_accessor).
- [ ] Create the has-many/belongs-to relationship between Classroom and Student. The following should be implemented:
- [ ] Create the has-many side (a classroom has many students).
- [ ] Create the belongs-to side (a student belongs to a classroom).
- [ ] Make sure that when adding a student to a classroom it also sets the classroom for the student.
- [ ] Make sure that when setting the classroom for a student it also adds it to the classrooms' students.
- [ ] Create a class Book with the following:
- [ ] @title and @author instance variables, should be initialized in the constructor.
- [ ] Setters and getters for instance variables (remember about attr_accessor).
- [ ] Create a class Rental with the following:
- [ ] @date instance variable, should be initialized in the constructor.
- [ ] Setter and getter for @date (remember about attr_accessor).
- [ ] Create the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental. The following should be implemented:
- [ ] Create the has-many side of Book and Rental (a book has many rentals).
- [ ] Create the belongs-to side of Rental and Book (a rental belongs to a book).
- [ ] Create the has-many side of Person and Rental (a person has many rentals).
- [ ] Create the belongs-to side of Rental and Person (a rental belongs to a person).
- [ ] Modify the constructor of Rental so Book and Person are set in it.

### CODE REVIEW

File: book.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+require_relative './rental'
+
+class Book
+  attr_accessor :title, :author
+  attr_reader :rentals
```
Comment (rloterh) [inline book.rb]: - [ ] `rental` should be marked with `attr_accessor`, as we intend to both _access (get)_ as well as _write (set)_ it. Kindly ensure to improve upon this aspect.

File: book.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def add_rental(rental)
+    @rentals << rental
+    rental.book = self unless book.rental == self
+  end
```
Comment (rloterh) [inline book.rb]: - [ ] Nice job so far creating the `add_rental` method, however, the _book_ itself should be sent as a parameter to create the new _rental_, thereby achieving the _'has-many'_ association. We can achieve this by modifying our code to look like the following example below:

```ruby
class Book
  attr_accessor :title, :author, :rentals
  .
  .
  def add_rental(person, date)
    Rental.new(date, self, person)
  end
end
```

File: rental.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+class Rental
+  attr_accessor :date
+  attr_reader :book, :person
```
Comment (rloterh) [inline rental.rb]: - [ ] Let's mark `book` and `person` with `attr_accessor`, as we intend to both _access (get)_ as well as _write (set)_ to it. Kindly ensure to improve upon this aspect.

File: person_class.rb [hunk 4]
```diff
@@ -2,16 +2,18 @@
 require_relative './base_decorator'
 require_relative './trimer_decorator'
 require_relative './capitalize_decorate'
+require_relative './rental'
 
 class Person < Nameable
   attr_accessor :name, :age, :parent_permission
-  attr_reader :id
+  attr_reader :id, :rentals
```
Comment (rloterh) [inline person_class.rb]: - [ ] Again, `rental` should be marked with `attr_accessor`, as we intend to both _access (get)_ as well as _write (set)_ it. Kindly ensure to improve upon this aspect.

File: person_class.rb [hunk 5]
```diff
@@ -23,6 +25,16 @@ def correct_name
     @name
   end
 
+  def add_rentals(rental)
+    @rentals << rental
+    rental.person = self
+  end
+
```
Comment (rloterh) [inline person_class.rb]: - [ ] Here again, let's ensure the _person_ itself is sent as a parameter to create the new _rental_, thereby achieving the _'has-many'_ association. We can achieve this by modifying our code to look like the following example below:

```ruby
class Person
  attr_accessor :name, :age, :rentals
  .
  .

  def add_rental(book, date)
    Rental.new(date, book, self)
  end

  .
  .
end
 ```

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Tracy, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project but you are almost there!
🏁 🚦 ⛪ 🚜 🏃 🌴 🏫 🌳 🏘️

## Highlights 👍  
- Good job so far with creating associations among objects!

## Required Changes ♻️
_Kindly check the inline comments under the review._

Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: rental.rb (hunk 3 above)
Comment (Tracy-miranja) [reply rental.rb]: @rloterh  kindly confirm what you are telling me to change is it right?

Comment (rloterh) [reply rental.rb]: @Tracy-miranja 
Yep, let change the encapsulation of _book_ and _person_ to attr_accessor, like below:
```ruby
class Rental
  attr_accessor :date, :book, :person
  ...
end
```

Cheers! 😃

File: book.rb (hunk 1 above)
Comment (peterrobert) [reply book.rb]: This change has not been implemented

File: book.rb (hunk 2 above)
Comment (peterrobert) [reply book.rb]: This change has not been implemented

File: rental.rb (hunk 3 above)
Comment (peterrobert) [reply rental.rb]: This change has been done well

File: person_class.rb (hunk 4 above)
Comment (peterrobert) [reply person_class.rb]: This change has been done well

File: person_class.rb (hunk 5 above)
Comment (peterrobert) [reply person_class.rb]: This change has been done well

File: book.rb (hunk 1 above)
Comment (peterrobert) [inline book.rb]: - [ ] Just like the first reviewer has said, rental should be marked with` attr_accessor`, as we intend to both access (get) as well as write (set) it. Kindly ensure to improve upon this aspect. It seems you missed implementing this change. Kindly do so

File: book.rb (hunk 2 above)
Comment (peterrobert) [inline book.rb]: - [ ] Just like the first reviewer has said, nice job implementing the add_rental method, however, the book itself should be sent as a parameter to create the new rental, thereby achieving the 'has-many' association. We can achieve this by modifying our code to look like the following example below:
```
class Book
  attr_accessor :title, :author, :rentals
  .
  .
  def add_rental(person, date)
    Rental.new(date, self, person)
  end
end
```

**Review by peterrobert — CHANGES_REQUESTED** [review]
Hi @Tracy-miranja , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏


Feel free to leave any questions or comments in the PR thread or tag me @peterrobert if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have only 3 code reviews per project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by binhussen — APPROVED** [review]
Hi @Tracy-miranja , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉

## Highlights
- Fixed Request Changes ✔️
- Set up associations✔️
- Used Git flow ✔️
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**